Email verification

remioluwatomi opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
​
## Description
<!--- Describe your changes in detail -->
This PR ensures that users must have a verified email before performing certain actions by enforcing email verification checks in middleware. It also updates test cases to properly handle email verification when creating users.  

​
## Related Issue (Link to Github issue)
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
Feat #620 
​
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Additional Security and Authentication Layer -> This PR addresses that by ensuring users have the `email_verified_at` field set. Additionally, it reinforces middleware checks for verified users.  

​
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Created test users with `email_verified_at` set.  
- Ran the full test suite to confirm that previously failing tests now pass.  
- Ensured that authentication and authorization behaviors remain unchanged apart from the intended verification requirement.
-Modified previous test to factor in the new Email Verification system. This test include:
    - InvitationTest.php
    - UserDeletionTest.php
​
<!--- ## Screenshots (if appropriate - Postman, etc): -->
​
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
​
## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

remioluwatomi committed:
feat: Add email verification for new users

Implement Laravel's MustVerifyEmail, verification routes, and protected middleware to ensure users verify their email before accessing protected routes. Improves security and prevents fake sign-ups.
Implement email.verified middleware (EnsureEmailIsVerified Middleware)
Updated the test to factor in email verification

remioluwatomi committed:
Merge branch 'dev' into email-verification
syncing local branch with remote upstream repo

bhimbho commented on app/Http/Middleware/EnsureEmailIsVerified.php:
```diff
@@ -0,0 +1,31 @@
+    public function handle(Request $request, Closure $next): Response
+    {
+        // Ensure user is authenticated
+        dump($request->user);
```
not great

bhimbho commented on app/Http/Controllers/Api/V1/Auth/VerificationController.php:
```diff
@@ -0,0 +1,59 @@
+    //
+
+
+    public function emailNotice($request)
```
incomplete logic

samuelogboye commented on app/Http/Controllers/Api/V1/Auth/VerificationController.php:
```diff
@@ -0,0 +1,59 @@
+    //
+
+
+    public function emailNotice($request)
```
in addition, what is emailNotice, notice for what?

bhimbho replied:
its a thing with laravel mustVerifyEmail trait. its a fallback

samuelogboye commented on tests/Feature/InvitationTest.php:
```diff
@@ -29,6 +29,11 @@ public function setUp(): void
             'last_name' => 'User',
         ]);
 
+        //pseudo email verification
+        $user = User::where('email', 'test@example.com')->first();
+        $user->email_verified_at = now();
+        $user->save();
+
```
Include more test edge cases, such as expired verification links or invalid tokens.

remioluwatomi replied:
Laravel requires three routes for email verification - according to the docs

- a notice route - This route is used to notify user that they have an unverified email. Also the route that laravels calls when they try to access a protected endpoint with an unverified email.

- verify email route -  This is the route that implements the actual verification of the email

- resend email route - this is needed incase they don't receive the original email or they misplace it

the notice route is required by laravel - i named the handler emailNotice.

you can check the official laravel docs: https://laravel.com/docs/11.x/verification

remioluwatomi committed:
fix: handle InvalidSignatureException in API responses

- Catch InvalidSignatureException - to check for Invalid/Expired Verification Link
- Ensure expired or invalid verification links return a proper JSON response
